### PR TITLE
Feature/pad

### PR DESCRIPTION
**Pull request by sbodenstein**
A padding layer `mx.Pad` that supports flexible padding: 

- Each tensor dimension can be padded at the start and end (but only special case of 2d images is supported right now, ie padding only the last two dimensions of a 4-tensor).
- Two padding modes are supported at the moment: `"constant"` and `"replicate"`. `"constant"` pads with a constant value, 0 by default, with a custom value specifiable with the `padding_constant` option. `mx.Pad` with `"constant"` pad mode is equivalent to [tf.pad](https://www.tensorflow.org/versions/r0.11/api_docs/python/array_ops.html#pad) with `mode='CONSTANT'`. `"replicate"` pads with the tensor boundary values, and is equivalent Torch [SpatialReplicationPadding](https://github.com/torch/nn/blob/master/doc/convolution.md#nn.SpatialReplicationPadding) for the special case of image padding. The implementation was adapted from Torch.

A general implementation will need to be added later.

Also note: `atomicAdd` was overloaded for `double` and `fp16` types. The latter needs to be carefully checked.

### CODE REVIEW

**Comment by sbodenstein** [comment]
@piiswrong

**Comment by sxjscience** [comment]
It seems more suitable to move the implementation of padding to the MShadow side. What do you think? @sbodenstein @piiswrong

**Comment by piiswrong** [comment]
It's hard to do in mshadow due to atomicadd.
Plus our guideline now is to avoid using mshadow for heavy ops. So this is good

@sbodenstein Could you add unittest?

**Comment by sbodenstein** [comment]
@piiswrong: I have done some renaming. I think following [numpy.pad](https://docs.scipy.org/doc/numpy/reference/generated/numpy.pad.html) will be the least confusing to users. Let me know if you are happy with this changed design.

Ok, will add some tests.

File: src/operator/pad-inl.h [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+        .add_enum("edge", pad_enum::kEdge)
+        .describe(
+            "Padding type to use. \"constant\" pads all values with a constant "
+            "value,"
+            "the value of which can be specified with the constant_value"
+            "option. \"edge\" uses the boundary values of the array as "
+            "padding.");
+
+    DMLC_DECLARE_FIELD(pad_width).describe(
+        "A tuple of padding sizes of length 2*r, where r is the rank of the "
```
Comment (piiswrong) [inline src/operator/pad-inl.h]: This doesn't really explain the format of the tuple and it's actually different from numpy.pad, which uses nested tuples

Comment (piiswrong) [inline src/operator/pad-inl.h]: I thinking moving it to pad_width is a good idea.
maybe additionally note its relation to pad_width

Comment (sbodenstein) [inline src/operator/pad-inl.h]: Ok, I've improved the docs, and mentioned the relationship between MXNet pad_width and Numpy pad_width.

**Comment by sbodenstein** [comment]
@piiswrong: how do I avoid the following lint errors:

```
04:38:28 src/common/cuda_utils.h:161:  Use int16/int64/etc, rather than the C type long  [runtime/int] [4]
04:38:28 src/common/cuda_utils.h:162:  Use int16/int64/etc, rather than the C type long  [runtime/int] [4]
04:38:28 src/common/cuda_utils.h:163:  Use int16/int64/etc, rather than the C type long  [runtime/int] [4]
04:38:28 src/common/cuda_utils.h:164:  Use int16/int64/etc, rather than the C type long  [runtime/int] [4]
```
If I follow this advice, then I will get build errors, as `atomicCAS` requires these C types.

**Comment by piiswrong** [comment]
you can disable lint for that line with // NOLINT(*)

**Comment by sbodenstein** [comment]
@piiswrong: for some reason the Scala test is broken, despite me changing no Scala code. It gives error:

> No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.

**Pull request merged**